testplan/exporters/testing/xml/__init__.py

```diff
@@ -280,5 +280,6 @@ def export(self, source):
                 )
 
         TESTPLAN_LOGGER.exporter_info(
-            '%s XML files created at: %s', len(source), xml_dir)
+            '{} XML files created at {}', len(source), os.path.abspath(xml_dir)
```

Comment:
Replacing %s with {} won't work, you have to use %-style formatting with python logging when the format values are specified as *args since it uses the % operator to do the format: https://docs.python.org/3/library/logging.html#logging.Logger.debug

Generally with logging it is slightly preferable to use % formatting and pass the format options as *args rather than formatting explicitly using either % operator or .format(). This is because the logging module can choose not to perform the string formatting at all if e.g. for a debug log if debug logging is not enabled, as an optimization. Whereas when you format a log message with .format() the string formatting always has to be done, even if the message is then discarded. The real impact of performing a few unnecessary string formats is almost certainly negligible, but it's just a nice detail to be aware of.


In any case, putting {} here without a .format will cause an Exception to be raised, so you need to fix - simplest to just revert back to %s flags.

Logging is just a special case in this regard - outside of logging, remember {} and .format are definitely preferable to % formatting!

Reply:
I saw **PDFExporter** also use "{}" in **TESTPLAN_LOGGER.exporter_info**, so maybe we also need to change that. Another thing is that **JSONExporter** use **self.logger.exporter_info** instead of **TESTPLAN_LOGGER.exporter_info**, is it better? if we use **self.logger.exporter_info** then we also prefer %s notation?

Reply:
Currently the code uses a mix of {} and % notation for logging in different parts. I personally prefer % in logging for the optimization reasons noted above. For all other string formatting (e.g. exception messages, print() calls) {} formatting is preferred.

It is better to use self.logger if you can, since in the log file that will give information about which class is making the log. TESTPLAN_LOGGER is only meant for code that can't use self.logger (i.e. outside of a class).

Reply:
As you've seen, there is a lot of code that doesn't follow the above guidelines. I don't care strongly enough to go through and fix up everything at once, I've mostly just been fixing up the logging in code that I touch.

Reply:
that makes sense. so for all exporters I removedd **TESTPLAN_LOGGER**, and use %s notation instead of '{}'.format(...).

By the way, there's another small change in this commit, **HTTPExporter::_upload_report** accepts a json data as argument instead of a TestReport instance, because the MS internal exporter has its own TestReport class and corresponding schema.